src/ops/base.rs

```diff
@@ -67,8 +249,46 @@ impl Ingredient for Base {
                 rs: Records,
                 _: &DomainNodes,
                 state: &StateMap)
-                -> Records {
-        rs.into_iter()
+                -> Option<Records> {
+        // Write incoming records to log before processing them if we are a durable node.
+        let records_to_return;
+        let mut copy_rs = rs.clone();
+        match self.durability {
+            BaseDurabilityLevel::Buffered => {
+                // Perform a synchronous flush if one of the following conditions are met:
+                //
+                // 1. Enough time has passed since the last time we flushed.
+                // 2. Our buffer of write records reaches capacity.
+                let num_buffered_writes = self.buffered_writes.as_ref().unwrap().len();
+                let has_reached_capacity = num_buffered_writes + rs.len() >= BUFFERED_WRITES_CAPACITY;
+                let elapsed = self.last_flushed_at.unwrap().elapsed();
+                let has_reached_time_limit = elapsed >= Duration::from_millis(BUFFERED_WRITES_FLUSH_INTERVAL_MS);
+
+                if has_reached_capacity || has_reached_time_limit {
+                    self.buffered_writes.as_mut().unwrap().append(copy_rs.as_mut());
```

Comment:
`Vec::append` expects a mutable reference, and what we get in the arg list is a vanilla `rs: Records`. If I instead pass rs as a `&mut` (`self.buffered_writes.as_mut().unwrap().append(&mut rs)`), I get a  "cannot borrow mutably" compilation error.

Once I changed `::on_input` to expect a mutable rs instead, then it all worked without the cloning.